mage_ai/orchestration/pipeline_scheduler.py

```diff
@@ -1397,106 +1411,154 @@ def schedule_all():
-                        pipeline_scheduler.logger,
-                        pipeline_scheduler.build_tags(),
+                        pipeline_run = create_and_cancel_pipeline_run(
+                            pipeline,
+                            pipeline_schedule,
+                            payload,
+                            message='Pipeline run limit reached... skipping this run',
+                        )
+                    else:
+                        pipeline_run = PipelineRun.create(**payload, create_block_runs=False)
+                        # Log Git sync status for new pipeline runs if a git sync result exists
+                        if git_sync_result:
+                            pipeline_scheduler = PipelineScheduler(pipeline_run)
+                            log_git_sync(
+                                git_sync_result,
+                                pipeline_scheduler.logger,
+                                pipeline_scheduler.build_tags(),
+                            )
+                        initial_pipeline_runs.append(pipeline_run)
+
+                # Enforce pipeline concurrency limit
+                pipeline_run_quota = None
+                if concurrency_config.pipeline_run_limit is not None:
+                    pipeline_run_quota = concurrency_config.pipeline_run_limit - \
+                        len(running_pipeline_runs)
+
+                if pipeline_run_quota is None:
+                    pipeline_run_quota = len(initial_pipeline_runs)
+
+                if pipeline_run_quota > 0:
+                    initial_pipeline_runs.sort(key=lambda x: x.execution_date)
+                    pipeline_runs_to_start.extend(initial_pipeline_runs[:pipeline_run_quota])
+                    pipeline_runs_excluded_by_limit.extend(
+                        initial_pipeline_runs[pipeline_run_quota:]
                     )
-                initial_pipeline_runs.append(pipeline_run)
-
-        # Enforce pipeline concurrency limit
-        pipeline_run_quota = len(initial_pipeline_runs)
-        if concurrency_config.pipeline_run_limit:
-            pipeline_run_quota = concurrency_config.pipeline_run_limit - \
-                len(running_pipeline_runs)
+            finally:
+                lock.release_lock(lock_key)
 
-        if pipeline_run_quota > 0:
-            initial_pipeline_runs.sort(key=lambda x: x.execution_date)
-            for r in initial_pipeline_runs[:pipeline_run_quota]:
-                PipelineScheduler(r).start(should_schedule=False)
+        pipeline_run_limit = concurrency_config.pipeline_run_limit_all_triggers
+        if pipeline_run_limit is not None:
+            pipeline_quota = pipeline_run_limit - len(
+                pipeline_runs_by_pipeline.get(pipeline_uuid, [])
+            )
+        else:
+            pipeline_quota = None
+
+        quota_filtered_runs = pipeline_runs_to_start
+        if pipeline_quota is not None and pipeline_quota >= 0:
```

Comment:
actually, i think we should set pipeline_quota to 0 if it's a negative number.
otherwise, if it's a negative number, we start ll the pipeline runs, which is not right.

Reply:
good point, let me also add a unit test for that